Do not depend on ocl-icd

jaimergp opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [X] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [X] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [X] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

After conversations had in https://github.com/openmm/openmm/issues/2401, we would like to make this decision and remove `ocl-icd` as a dependency because it has the potential to degrade OpenMM performance. 

While OpenMM supports OpenCL execution (in addition to CUDA), OpenMM also provides its own CPU-optimized implementation to fall back on should neither OpenCL nor CUDA be available. The OpenCL implementation is intended to achieve high performance on vendor-provided optimized OpenCL implementations (GPU, mainly) that should be available if the corresponding hardware + drivers are also available, but we feel that---at this time---the fall-back OpenMM-provided CPU platform should be used if vendor-provided OpenCL implementations are not available.

Given the way `openmm` recognizes and loads platforms, adding `ocl-icd` would also mean that an OpenCL CPU implementation would have precedence over the builtin `CPU` platform. Since `openmm` has only tested---and currently only offers validated support for--- Nvidia, AMD and Intel OpenCL implementations (none of them installable through `conda-forge`), this could result in untested installations that might end up providing wrong results.

As `openmm` knows how to discover the system OpenCL installation on its own, there is no technical reason or need to require `ocl-icd`. It would only offer a potential source of errors and maintenance burden. Also, if a different OpenCL implementation needs to be used, the user can always install `ocl-icd` together with `pocl` or similar libraries. We are open to test the performance of this approach in the future.

We hope our vision can fit in the wonderful `conda-forge` ecosystem, for which we are thankful and delighted to be a part of!

(Pinging @isuruf and @jakirkham for feedback).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [40]

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

isuruf commented:
I'm -1 to this. conda-forge is not packaging one software. It is a software stack and packages need to play with one another nicely. You are thinking about one package while we need to think about the stack of packages.

Your suggestion to not use `libOpenCL.so` from conda-forge is like asking to use `libz.so` from system because almost all distributions have them.

One bad interaction would be when you load a slightly older `libOpenCL.so` from the system which supports only OpenCL 1.2 (For example, the NVIDIA provided `libOpenCL.so` from cuda driver 340. I know this is ancient, but this just an example) and a user imports `openmm` first and then tries to import another package needing OpenCL 2.0 (For eg: pyopencl from conda-forge) which is linked to the conda library. Since their sonames are the same and the older library is already loaded into the process, the conda library will not be loaded, which would result in missing symbols.

`openmm` in isolation would work correctly, but that would mean you'll be breaking other software.

jchodera commented:
> Btw, Intel OpenCL NEO is available in conda-forge as intel-compute-runtime.

Cool! We'll add this to our set of OpenCL implementations to test!

jakirkham commented:
I'm curious what issues are these packages still causing? It was my understanding that the `ocl-icd-system` shim package solved the issues of ensuring the system OpenCL was used. Were there other issues that have not yet been addressed? If so, could you please share them?

jchodera commented:
> I'm curious what issues are these packages still causing? It was my understanding that the ocl-icd-system shim package solved the issues of ensuring the system OpenCL was used. Were there other issues that have not yet been addressed? If so, could you please share them?

Here's the primary set of issues:
* If the `ocl-icd-system` shim package is installed, then there is no way that OpenMM will fall back to its optimized `CPU` platform since it will always find an OpenCL implementation. We _want_ it to fall back to our optimized `CPU` implementation if it doesn't have hardware OpenCL installed.
* Whereas we _have_ extensively tested our `CPU` implementation on a ton of hardware, we haven't tested some of the OpenCL implementations that `ocl-icd-system` would provide if the optimized vendor OpenCL wasn't available. So it could produce wrong results.

We have not identified _any_ disadvantages to removing the dependency on `ocl-icd-system`.

> I'm -1 to this. conda-forge is not packaging one software. It is a software stack and packages need to play with one another nicely. You are thinking about one package while we need to think about the stack of packages.

Absolutely! We want everything to play nicely with each other.

> Your suggestion to not use libOpenCL.so from conda-forge is like asking to use libz.so from system because almost all distributions have them.

I'm not quite sure that's correct. OpenMM supports a variety of hardware acceleration implementations if they are available, including CUDA. We also have an optimized `CPU` implementation platform that we want to use if there are no hardware-accelerated OpenCL devices available (as noted above).

> One bad interaction would be when you load a slightly older libOpenCL.so from the system which supports only OpenCL 1.2 (For example, the NVIDIA provided libOpenCL.so from cuda driver 340. I know this is ancient, but this just an example) and a user imports openmm first and then tries to import another package needing OpenCL 2.0 (For eg: pyopencl from conda-forge) which is linked to the conda library. Since their sonames are the same and the older library is already loaded into the process, the conda library will not be loaded, which would result in missing symbols.

I believe we fully support OpenCL 1.2, so this would not be a problem for OpenMM.

> openmm in isolation would work correctly, but that would mean you'll be breaking other software.

I don't quite understand how we would _break_ other software, but I'm very eager to learn how we can avoid this, since this is not our intention! 

As I understand it, if _another_ package wants to install `ocl-icd-system`, then that should be totally fine. It _will_ cause OpenMM to prefer whatever OpenCL is provided over the `CPU` platform, so we will want to start testing this soon after our release. But I think the only risk here is that installation of another package that bings in `ocl-icd-system` could potentially cause OpenMM to produce incorrect results under some circumstances because we have not yet performed testing on some of the OpenCL implementations it provides---but I don't see the risk of OpenMM breaking _other_ packages.

If I'm misunderstanding this, my apologies, and I'd ask your patience in explaining the specific risk to other packages and how we can avoid it (while also not forcing OpenMM to potentially use an untested OpenCL implementation by default).

isuruf commented:
> I believe we fully support OpenCL 1.2, so this would not be a problem for OpenMM.

I don't get what you are saying here. It's not that openmm support opencl 1.2. Issue is that OpenMM might link to a system libOpenCL.so that does not support OpenCL 2.0 and a conda package requiring OpenCL 2.0 will not work.

isuruf commented:
> If the ocl-icd-system shim package is installed, then there is no way that OpenMM will fall back to its optimized CPU platform since it will always find an OpenCL implementation. 

If `ocl-icd-system` is installed and there's no conda OpenCL package nor any system OpenCL, how can it find an OpenCL implementation?

jchodera commented:
> If ocl-icd-system is installed and there's no conda OpenCL package nor any system OpenCL, how can it find an OpenCL implementation?

Ah, I see! So `ocl-icd-system` does not bring in any OpenCL implementations by default?

> I don't get what you are saying here. It's not that openmm support opencl 1.2. Issue is that OpenMM might link to a system libOpenCL.so that does not support OpenCL 2.0 and a conda package requiring OpenCL 2.0 will not work.

I think I'd misunderstood your earlier comment. I went back and read https://github.com/openmm/openmm/issues/2401#issuecomment-540142993 and think I better understand the role of `ocl-icd-system`. I'm just having a difficult time understanding the specific concern about how OpenMM could _break_ other packages that use `ocl-icd-system`---is the concern regarding importing two conda-forge packages that link to different `libOpenCL.so` into the same interpreter? If so, then I agree with you---this could be problematic and we should reconsider.

isuruf commented:
> So ocl-icd-system does not bring in any OpenCL implementations by default?

Yes, it just brings in the ability to see OpenCL implementations in both `/etc/OpenCL/vendors` and `$CONDA_PREFIX/etc/OpenCL/vendors`. No OpenCL implementations are brought in by default. `ocl_icd_wrapper_apple` brings in the Apple implementation by default on OSX, but that's expected.

> is the concern regarding importing two conda-forge packages that link to different libOpenCL.so into the same interpreter? If so, then I agree with you---this could be problematic and we should reconsider.

Yes, the way the loader works is that two different libOpenCL.so will not be loaded. Only the first one will get loaded. If the first one is the newer one, yay. If not, you get crashes. This is why the OpenCL standard says that a vendor should not overwrite a `libOpenCL.so` with an older vendor specific one.

BTW, you can have libOpenCL.so and a CPU implementation installed systemwide (That's what I have in machines without GPUs) and then `openmm` will prefer the OpenCL CPU over openmm's CPU platform. So, this problem of preferring OpenCL CPU platforms is not unique to conda-forge packaging.

jchodera commented:
OK, I think I concur that we _should_ be using `ocl-icd-system` to avoid issues when using OpenMM with other packages that use OpenCL.

peastman commented:
Let me state this clearly and emphatically, because I've said it several times already and you don't seem to be hearing it:

*There is precisely no reason whatsoever for OpenMM to list ocl-icd as a dependency.*

OpenMM does not require it.  It works just fine without it.  If it is installed, it *might* cause OpenMM to work incorrectly.  In our next release we will consider adding checks to explicitly block unsupported OpenCL implementations.  But regardless of that, there simply is zero benefit to installing it.

If a user happens to have installed it, so be it.  We'll try to work around that and make OpenMM work correctly in that case.  But installing OpenMM should not *cause* ocl-icd to be installed.

jchodera commented:
@peastman: Can we have a brief developer huddle? You can call me now if you have time (phone number in my email signature).

peastman commented:
By the way, I have a suggestion.  Rather than making ocl-icd a dependency of packages that *use* OpenCL, wouldn't it make more sense for it to be a dependency of ones that *implement* OpenCL?  So for example, if a user installs pocl that would cause ocl-icd to be installed as well, thus ensuring it can be loaded.

A package that uses OpenCL might then choose to list pocl as a dependency.  Or it could not list it, and then it would only work if the user had a standard implementation already installed.  ocl-icd would be installed if and only if it was needed to let a conda-installed OpenCL implementation be loaded.

isuruf commented:
> There is precisely no reason whatsoever for OpenMM to list ocl-icd as a dependency.

>  But regardless of that, there simply is zero benefit to installing it.

Please don't make claims such as this without reading the whole thread.

peastman commented:
> Please don't make claims such as this without reading the whole thread.

I have, in fact, read the whole thread.  What do you believe is the benefit to making it a dependency?

jchodera commented:
@isaruf: @jaimergp and I have been working on testing this further, and we have now empirical evidence that the current conda-forge build of OpenMM is defective because of the inclusion of `ocl-icd` and `ocl-icd-system`. See this PR: https://github.com/choderalab/openmmtools/pull/442

I'd love to better understand what might be going on, and why including `ocl-icd` and `ocl-icd-system` might be causing OpenMM to try to use the OpenCL platform when it should not (because no OpenCL implementation should be installed) rather than fall back to the CPU platform. Do you have any insight here?

isuruf commented:
`Error initializing context: clGetPlatformIDs (-1001)` happens when you try to create an opencl context without any implementation. I guess `openmm` relies on the fact that if an OpenCL loader is found, there is an implementation. This is an issue with `openmm`. This is not unique to `conda` and can be recreated in linux distros like debian too.

jchodera commented:
@isuruf: Is there a canonical way of safely querying whether an implementation is available once a loader is found? We might be able to get this into the forthcoming OpenMM bugfix release. 
(Apologies if this is a naïve question---I only contribute to the Python part of OpenMM development.)

isuruf commented:
Sure. See https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clGetPlatformIDs.html

isuruf commented:
Note that you should check that both CL_SUCCESS is returned and `num_platforms` returned is not 0.

jchodera commented:
@jaimergp : Now that https://github.com/openmm/openmm/pull/2441 is merged to address https://github.com/conda-forge/openmm-feedstock/pull/7#issuecomment-544194779, let's try basing this test build on OpenMM commit [`7a02c59b7f94c26c7996c2a0c4ef98d5fa4af7cd`](https://github.com/openmm/openmm/commit/7a02c59b7f94c26c7996c2a0c4ef98d5fa4af7cd) and do some further testing of the conda package without `ocl-icd` once its merged. 

We'll still keep pushing to the `testing` label for now.

We might also consider testing _with_ `ocl-icd` in a separate label so we can determine which variant to pursue for the upcoming OpenMM 7.4.1 release.

jaimergp commented:
@peastman 

In this last commit:

- I am building from commit `7a02c59b7f94c26c7996c2a0c4ef98d5fa4af7cd`, which includes the OpenCL fixes
- Added more tests to see how we handle `ocl-icd` (or its absence):
  - In Linux, we should only see two platforms with or without `ocl-icd` (CPU and Reference). If we additionally install `ocl-icd` + `pocl`, then we should see OpenCL too. However, we cannot test this because right now `import simtk.openmm` fails with `clGetPlatformIDs`
  - In MacOS, `openmm` should be able to see the system OpenCL without `ocl-icd`, but this is not happening.

Any ideas?

isuruf commented:
1 is because that commit only checks that the number of platforms is 0, but throws an exception when `clGetPlatformIds` doesn't return success. See the description in https://github.com/openmm/openmm/issues/2433#issue-509526987

isuruf commented:
Note that you run into this problem even without `ocl-icd` because there's libOpenCL.so installed in the docker image, but there's no GPU in CI.

jchodera commented:
@jaimergp : OK, now that https://github.com/openmm/openmm/pull/2445 has been merged, let's try again with the latest commit!

isuruf commented:
MacOS issue is because you are linking to libOpenCL.dylib at build time. This is an ICD loader, but Apple's system implementation is a framework library with no ICD interface. (We work around this by using the ICD loader and then having a small wrapper library `ocl_icd_wrapper_apple` that exposes the Apple framework as an ICD) If you are removing ocl-icd, you should link with the framework.

peastman commented:
Could you clarify why it's failing?  The log reports an exception in `clGetPlatformIDs()`.  But that function should work fine with Apple's OpenCL framework.

isuruf commented:
> The log reports an exception in clGetPlatformIDs()

That's on Linux.

peastman commented:
So what is the cause of the Mac error?  It's coming from the line

```
python -c 'from simtk.openmm import Platform as P; n = P.getNumPlatforms(); assert n == 3, f'\''# platforms ({n}) != 3'\'''
```

It's expecting there to be three registered platforms, but there are only two.  The OpenCL framework is built into the OS, and it includes a CPU implementation that we don't support but that should still guarantee that at least one implementation is always found.

jaimergp commented:
I have added some more tests to check the compatibility with `ocl-icd`. Seems to be working ok now after the last fix in `openmm`.

I would like to move this forward, so let me sum up the current status of the conversation:

__With `ocl-icd`__

* There were concerns that an unsupported OpenCL implementation would end up being loaded, in the worst case leading to a functional installation with wrong results. However, this can happen regardless _we_ install `ocl-icd`: if another OpenCL implementation is installed, then `ocl-icd` will be installed anyway.
* Before commit `8ac164930170af8ff7460e448abb721b3ba526e7`, `openmm` did not check correctly for existing implementations and having `ocl-icd` would trigger a platform error. This is solved now.
* The only requirement is that `ocl-icd-system` must be installed to save the user from manually executing `ln`.

__Without `ocl-icd`__

* There are concerns that an outdated OpenCL implementation installed system-wide can be loaded in the Python process before the implementation installed in the `conda` environment. However, if the environment contains an OpenCL implementation, `ocl-icd` should be [*] there anyways, so I don't see how this scenario is feasible.
* Additionally, if we use `ocl-icd-system` to symlink the system-wide installation, wouldn't this be the same problem? Or does `ocl-icd` have a smart mechanism to deliver the needed OpenCL implementation to different libraries in the Python process?

[*] By the way, I tried `intel-compute-runtime` and was surprised to see that `ocl-icd` is not listed in the runtime requirements. Shouldn't at least all OpenCL implementations require this?

<details>
<summary>

`conda create -n intel intel-compute-runtime`

</summary>

```
Collecting package metadata (current_repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.7.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base conda



## Package Plan ##

  environment location: /home/jaime/.conda/envs/intel

  added / updated specs:
    - intel-compute-runtime


The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main
  intel-compute-run~ conda-forge/linux-64::intel-compute-runtime-19.39.14278-hc9558a2_0
  intel-gmmlib       conda-forge/linux-64::intel-gmmlib-19.3.2-hc9558a2_0
  intel-graphics-co~ conda-forge/linux-64::intel-graphics-compiler-1.0.2652-hc9558a2_0
  intel-opencl-clang conda-forge/linux-64::intel-opencl-clang-9.0.0-hc9558a2_0
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0
  libllvm9           conda-forge/linux-64::libllvm9-9.0.0-hc9558a2_2
  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0
  llvm-spirv         conda-forge/linux-64::llvm-spirv-9.0.0_1-hc9558a2_0


Proceed ([y]/n)?

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate intel
#
# To deactivate an active environment, use
#
#     $ conda deactivate
```

</details>

***

I honestly do not have a preference. Both options work. Not having `ocl-icd` installed less moving parts, and thanks to this discussion `openmm` knows how to deal with the ICD loaders.

isuruf commented:
Can you also add MacOS to the summary? Without the khronos ICD loader at build time, you are linking to the Apple framework instead of `libOpenCL.dylib`, which means installing the khronos ICD loader will not work with openmm even if installed later.

isuruf commented:
> By the way, I tried intel-compute-runtime and was surprised to see that ocl-icd is not listed in the runtime requirements. Shouldn't at least all OpenCL implementations require this?

No, some of these OpenCL implementations allow linking directly instead of using an ICD loader. So, the packages using OpenCL should bring in the ICD loader, not the OpenCL implementations.

isuruf commented:
> Or does ocl-icd have a smart mechanism to deliver the needed OpenCL implementation to different libraries in the Python process?

There is `OCL_ICD_VENDORS` env variable that lets you set the OpenCL implementation you want for all libraries and you can also use that env variable to hide all the OpenCL implementations. This might or might not be supported by the OpenCL implementation that is installed system wide.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [83]

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

peastman commented:
> There are concerns that an outdated OpenCL implementation installed system-wide can be loaded in the Python process before the implementation installed in the conda environment. 

Since the only OpenCL implementations we support are ones that get installed system wide, and we don't support any of the conda installable ones, this is the correct behavior.

> Without the khronos ICD loader at build time, you are linking to the Apple framework instead of libOpenCL.dylib, which means installing the khronos ICD loader will not work with openmm even if installed later.

Here again this is the desired behavior.  The only OpenCL we support on Mac OS is the one from Apple that's built into the OS.

isuruf commented:
> Here again this is the desired behavior. The only OpenCL we support on Mac OS is the one from Apple that's built into the OS.

If you do that and another package uses the OCL loader, you'll be bringing in two OpenCL implementations. (Apple framework and the OCL loader)

peastman commented:
> If you do that and another package uses the OCL loader, you'll be bringing in two OpenCL libraries. (Apple framework and the OCL loader)

I assume that would produce a runtime error?  Or would the two libraries just run independently, each using a different OpenCL?  Either way, that sounds fine.  Much better than risking using an unsupported OpenCL.

isuruf commented:
> I assume that would produce a runtime error? 

It would probably crash.

> Much better than risking using an unsupported OpenCL.

They'd both be using the same OpenCL implementation, but in two different ways. Openmm would link directly to Apple and the other would load it using an ICD loader.

Note that ocl-icd-system and the khronos loader does not bring in any OpenCL implementations by default. It will use only the implementations in the system. A user has to explicitly install an OpenCL implementation for it to be visible.

If you are against using unsupported OpenCL implementations, then please add code into OpenMM to avoid any unsupported implementations.

peastman commented:
> If you are against using unsupported OpenCL implementations, then please add code into OpenMM to avoid any unsupported implementations.

We'll certainly consider doing that in a future release.  But for the moment, let's just continue using Apple's OpenCL framework.  It's a standard system framework that's built into the OS, and it's what pretty much all OpenCL codes on macOS do.  If ocl-icd produces a crash when any other package links to a standard system framework, that's a bug in ocl-icd.

peastman commented:
@jaimergp is there something we can put in the recipe to indicate that OpenMM and ocl-icd are incompatible with each other, so conda won't let them be installed in the same environment?  That would be a good solution.  We're spending far too much time trying to deal with a situation that in practice has never come up, and quite possibly never will come up.

jchodera commented:
Formally, I don't believe there's a way to specify "this package can NOT be installed to permit installation of `openmm`", but practically, it seems we could abuse the [`Run_constrained`](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#run-constrained) option by specifying
```yaml
requirements:
  run_constrained:
    - olc-icd == 999.999.999
```
or some version that doesn't exist. 

However, I'd rather we try to move ahead without `ocl-icd` and then both (1) extensively test for any incompatibilities this may cause with other packages that use `ocl-icd`, and (2) build test variants with `ocl-icd` and see if these other OpenCL implementations can be validated with OpenMM. I think this would both satisfy our short-term needs and allow us to work toward making OpenMM fully compatible with all the GPU-accelerated packages in the `conda-forge` ecosystem.

Of course, if there are any major issues detected, we can work rapidly to address them after release.

@isuruf : Does this sound like a reasonable approach for now? We're running into a time crunch with the impending need to release OpenMM 7.4.1 on `conda-forge`.

jaimergp commented:
Following on the tests for ICD on/off, for MacOS:

If `openmm` is compiled without Khronos' ICD loaders, it will always find the system installation first, even if we manually install `khronos-opencl-icd-loader` and, let's say, `pocl`. Using the system implementation throws the following error:

```
OpenCL platform error: No compatible OpenCL platform is available
```

Is that due to no GPU being found and Apple's OpenCL not being able to use CPU devices?

Note that the Linux build is indeed able to use `ocl-icd` + `pocl` even if we link against Nvidia's OpenCL libraries.

peastman commented:
> Is that due to no GPU being found and Apple's OpenCL not being able to use CPU devices?

Correct.  Technically Apple's OpenCL does support CPUs, but it does it in such a limited way that it's useless for us, so we reject it.

jchodera commented:
Is everyone OK with moving forward with [this plan](https://github.com/conda-forge/openmm-feedstock/pull/7#issuecomment-546536916)?

peastman commented:
Ok with me.

jchodera commented:
@jaimergp : Want to merge in the changes from `master`, bump the build number, and merge?

isuruf commented:
> Is everyone OK with moving forward with this plan?

Nope.

jchodera commented:
> Nope.

@isuruf : Can you clarify further on the harm it would cause for moving forward with this plan for now? Here's a copy of what I said above proposing how we can move forward:

> However, I'd rather we try to move ahead without ocl-icd and then both (1) extensively test for any incompatibilities this may cause with other packages that use ocl-icd, and (2) build test variants with ocl-icd and see if these other OpenCL implementations can be validated with OpenMM. I think this would both satisfy our short-term needs and allow us to work toward making OpenMM fully compatible with all the GPU-accelerated packages in the conda-forge ecosystem.
>
> Of course, if there are any major issues detected, we can work rapidly to address them after release.
> 
> @isuruf : Does this sound like a reasonable approach for now? We're running into a time crunch with the impending need to release OpenMM 7.4.1 on conda-forge.

Again, we're happy to make changes quickly if we notice actual issues in the field or package incompatibilities. But we (1) need to get OpenMM 7.4.1 out, and as far as we know, this plan should work for all our users, and (2) we can continue to test after release to see how we might better convince ourselves we can include `ocl-icd` with no problems and the alternative OpenCL implementations will give valid results.

jchodera commented:
@jaimergp : We're cutting OpenMM 7.4.1 right now---check https://github.com/openmm/openmm/issues/2430

We can also update this PR to build the 7.4.1 release and push to `testing` for more testing.

@peastman: Is it OK if we continue to use the `testing` label until we're ready for release? I think release will have to rebuild and push to the `main` label since we don't have access to changing the labels.

peastman commented:
> I think release will have to rebuild and push to the main label since we don't have access to changing the labels.

We need some way to push release candidates to production without rebuilding.  As we know from frequent experience, simply repeating the build a few weeks later often leads to new problems, if any of the build time dependencies has been updating in the intervening time.

jchodera commented:
> We need some way to push release candidates to production without rebuilding. 

@jaimergp: Any idea if we can do this?

jchodera commented:
I suppose an alternative would be to pin the package versions installed during build time to make it fully reproducible.

peastman commented:
Even if we pin all the conda dependencies, is it guaranteed nothing else will have changed?  For example, that the VM won't have upgraded to a newer version of gcc or anything like that?

isuruf commented:
> Can you clarify further on the harm it would cause for moving forward with this plan for now?

Your tests fail on OSX. Also you are going against conda-forge policies. We require all packages to depend on conda packages or CDT/compiler libraries. (unwritten at the moment, because we haven't had any many exceptions to this) If you want to change this, you need to open an issue in https://github.com/conda-forge/conda-forge-enhancement-proposals

peastman commented:
> Also you are going against conda-forge policies.

Could you be more specific?  Are you saying conda-forge has a policy against using libraries that are built into the OS?

isuruf commented:
Yes, examples are `libstdc++.so, libgcc-ng.so, libc++.dylib, Accelerate framework`. Only packages we allow in Linux are CDT/compiler packages which include are system libraries like `libc.so, libm.so`, etc and CDT ones like `libGL.so, libcuda.so.1`, etc.

isuruf commented:
> For example, that the VM won't have upgraded to a newer version of gcc or anything like that?

FYI, `gcc`, `libgcc_s`, `libstdc++` comes from a conda package.

jchodera commented:
We should definitely address the osx build failures:
```
+ python benchmark.py --test=rf --seconds=10 --platform=OpenCL
Platform: OpenCL
Precision: single

Test: rf
Step Size: 2 fs
Traceback (most recent call last):
  File "benchmark.py", line 161, in <module>
    runOneTest(options.test, options)
  File "benchmark.py", line 111, in runOneTest
    context = mm.Context(system, integ, platform, properties)
  File "/usr/local/miniconda/conda-bld/openmm_1572014862000/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.6/site-packages/simtk/openmm/openmm.py", line 5557, in __init__
    _openmm.Context_swiginit(self, _openmm.new_Context(*args))
Exception: No compatible OpenCL platform is available
Tests failed for openmm-7.4.0-py36h45c9a53_7.tar.bz2 - moving package to /usr/local/miniconda/conda-bld/broken
```
This is presumably the desired behavior, so we should ensure our tests correctly ascertain no valid OpenCL platform is detected.

Regarding the GPU library dependencies, my understanding (consistent with the fact that these policies are not yet written down) is that there are still few GPU-enabled packages on `conda-forge` and that the practice here is still evolving. 

> If you want to change this, you need to open an issue in https://github.com/conda-forge/conda-forge-enhancement-proposals

None of these yet deal with the treatment of GPU-targeted code or CUDA/OpenCL dependencies, so we would presumably want to start a collective dialog about the best way to handle these situations going forward.

I imagine that the idea here is to ensure the broadest package compatibility and interoperability---a goal I fully support!---but not to make choices that could cause software used by the community (in this case, a scientific software tool with 244K conda downloads) to potentially produce silently scientifically invalid results. I'm 100% committed to working to ensure broad package compatibility, but it would seem imprudent to take steps to address compatibility problems we don't yet have evidence of if this could lead to scientifically invalid results without thorough testing first.

My proposal is to address the failing test issues, release openmm 7.4.1 on conda-forge for testing, and continue to work to see if we can satisfy all the desires here---both compatibility and testing-validated correctness---as we continue to move forward. I don't think that's an unreasonable thing to propose! 

Since there's a larger evolving discussion regarding best practices for GPU-accelerated code, we should continue to participate that discussion and hopefully find the best way to address all of these issues in a policy standard everyone can agree to.

jchodera commented:
(Also, I want to say that we very much appreciate the patience you've had and the guidance you've given us here! Thank you!)

isuruf commented:
> This is presumably the desired behavior, so we should ensure our tests correctly ascertain no valid OpenCL platform is detected.

How is it that in linux case if pocl is installed, that's a valid platform, but not in OSX?

> My proposal is to address the failing test issues, release openmm 7.4.1 on conda-forge for testing, and continue to work to see if we can satisfy all the desires here---both compatibility and testing-validated correctness---as we continue to move forward. I don't think that's an unreasonable thing to propose!

If it's only for the `testing` label, then I don't think it is unreasonable. If it's for `main`, then I do find it unreasonable although you can open an issue in https://github.com/conda-forge/conda-forge-enhancement-proposals to get the opinion of others.

jchodera commented:
Excellent points. I think ideally, we'd:
* ensure `linux` and `osx` behave identically in our 7.4.1 release candidate
* prevent anything but supported OpenCL implementations from being used in our release candidate
* build another testing version that uses `ocl-icd` and `ocl-icd-system` where we can allow extensive testing against various other OpenCL implementations and sort out how to select among them
* raise the issue in https://github.com/conda-forge/conda-forge-enhancement-proposals to discuss if it would be reasonable to release the 7.4.1 release candidate if it passes testing, and policies for how OpenCL implementations should be dealt with in general

peastman commented:
> How is it that in linux case if pocl is installed, that's a valid platform, but not in OSX?

Because of the difference in behavior you described between Apple OpenCL and ocl-icd.

> Only packages we allow in Linux are CDT/compiler packages which include are system libraries like libc.so, libm.so, etc and CDT ones like libGL.so, libcuda.so.1, etc.

What about other platforms?  Are Mac programs forbidden to use Cocoa frameworks?  Are Windows packages not allowed to use win32 APIs?

jchodera commented:
@peastman: What would you think about opening an issue in https://github.com/conda-forge/conda-forge-enhancement-proposals to discuss the issue of OpenCL libraries for conda-forge? I think @isuruf is right that it would be useful to get a broader discussion going about the best way to accomplish our shared goals in supporting OpenCL for high-performance scientific computing.

jchodera commented:
@peastman: The [tests are failing on `osx`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=84227&view=logs&jobId=9c5ef928-2cd6-52e5-dbe6-9d173a7d951b) in this PR right now because removing the `ocl-icd` package causes OpenMM to no longer be able to use conda-forge installed OpenCL implementations (here, `pocl`) if a user wants to optionally install them. On `linux`, however, OpenMM _will_ use `pocl` if the user installs it from conda-forge.

Idea: Instead of neglecting `ocl-icd` with the hope of using this as a mechanism of rejecting unsupported OpenCL implementations, is there some other way we can query each available OpenCL implementation to see if it satisfies our requirements? It looks like the strategy of simply omitting `ocl-icd` may not actually work for us, since `linux` will still happily use the conda-forge installed implementations.

peastman commented:
This just needs minor changes to the recipe.  Currently it assumes it can install pocl and that will work: https://github.com/conda-forge/openmm-feedstock/blob/43c84853a97603c819d884adcf104a5f20cafb48/recipe/meta.yaml#L77-L85.  On Mac OS it doesn't work and it shouldn't be assuming that.

jaimergp commented:
Oh, let me change it then. On Linux we shouldn't be able to use `pocl` either, right?

peastman commented:
At present I think that test will pass on Linux.  But it may not pass in the future.

jaimergp commented:
Thanks everybody for the useful discussions. Based on this input, we are working on a path towards distributing `openmm` both on `conda-forge` (following CF best practices; i.e. using OpenCL ICD loaders) and `omnia` (with more CUDA versions and other differences). I'll close this PR but further discussion on the topic can still happen on https://github.com/conda-forge/conda-forge-enhancement-proposals/issues/17. Thanks again!
